Explore: Sort order of log results

ivanahuckova opened the pull request:
**What this PR does / why we need it**:
[designs in Figma](https://www.figma.com/file/7ZHtMoaK4RxBz3b6m6U0cW/Show-more-logs-%2B-Sorting?node-id=68%3A291)

- Adds button to flip order of log results. I have added test coverage. 
- Flips the order of context, if flipped order of logs.
- Refactors the sorting in Logs Panel, to keep things consistent. 
- Moves SortOrder and sorting function from core to grafana/data, as we need to use it in grafana/ui.


![image](https://user-images.githubusercontent.com/30407135/88816652-e983b080-d1bc-11ea-84fc-5f15084ba4f4.png)

Before updating button:
![Kapture 2020-07-29 at 13 16 25](https://user-images.githubusercontent.com/30407135/88793946-f349eb80-d19d-11ea-99e2-602a917d449b.gif)

**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/grafana/issues/16555

**Special notes for your reviewer**:

ivanahuckova committed:
Create sorting button and functionality

ivanahuckova committed:
Set up logs ordering

ivanahuckova committed:
Add tests

ivanahuckova committed:
Refactor

ivanahuckova committed:
Refactor

ivanahuckova commented:
@jessover9000 I have cople of question regarding UI:

- I have used a new button component, but it is the only new button component on screen (all other are old). Do you think it is okay?
- Should we add any visual feedback that the lines are switching order (e.g. icon in button with arrow up/down)?
- Where should the flip button be positioned on smaller screens?
![image](https://user-images.githubusercontent.com/30407135/88793533-39527f80-d19d-11ea-91fa-a012c752797f.png)

jessover9000 commented:
> it is the only new button component on screen (all other are old)

Will it be a lot of work switching that around later if we keep the old one for now? I generally think we should try and get Explore up to date with the new components sooner than later so I wouldn't mind the new button, but you are right that it looks a bit off – mostly because the font size is bigger than in the other buttons + the "dedup" radio group elements. Have you tried the small button size?

> Should we add any visual feedback that the lines are switching order (e.g. icon in button with arrow up/down)?

Do you mean that the icon would show whether the flipping is on or off? Or whether the logs are currently ascending or descending? Unicons doesn't have super clear icons for "ascending" and "descending" and I also think on a button it could be confusing for the user – does the button tell me what I currently see, or does the button tell me what I will see after I clicked it? I thought we wanted to do just a button without a state exactly because it's so simple and just a one-time action.

Responsive behaviour: I think once the elements in that row start breaking up into lines, the button should get its own line and stick to the left side of the box.

ivanahuckova committed:
Replace new button with old

ivanahuckova commented:
> Will it be a lot of work switching that around later if we keep the old one for now? 

No problem at all, I switched to old style button - totally agree with everything you've said. 

>I thought we wanted to do just a button without a state exactly because it's so simple and just a one-time action.

Yes I agree. Makes sense. Just was checking.

>Responsive behaviour: I think once the elements in that row start breaking up into lines, the button should get its own line and stick to the left side of the box.

✅ Done
![image](https://user-images.githubusercontent.com/30407135/88816363-990c5300-d1bc-11ea-9d24-8affc1801707.png)
![image](https://user-images.githubusercontent.com/30407135/88816392-a32e5180-d1bc-11ea-8a6a-e09dccfe077a.png)

aocenas commented on packages/grafana-ui/src/components/Logs/LogRows.tsx:
```diff
@@ -23,6 +23,7 @@ export interface Props extends Themeable {
   showTime: boolean;
   wrapLogMessage: boolean;
   timeZone: TimeZone;
+  logsOrder?: any;
```
Seems like this could be of type `SortOrder`.

aocenas commented on public/app/features/explore/Logs.tsx:
```diff
@@ -251,6 +278,8 @@ export class Logs extends PureComponent<Props, State> {
           wrapLogMessage={wrapLogMessage}
           timeZone={timeZone}
           getFieldLinks={getFieldLinks}
+          orderLogs={sortLogRows}
```
Do we need to pass this as prop? Is that just because one is in package and one in core? Personally I think we could move it to grafana-data package.

aocenas commented:
Another question is what to do with log context? Should we flip order there too? Otherwise I feel it gets more confusing if context windows are ordered differently.

ivanahuckova commented:
>Another question is what to do with log context? Should we flip order there too? Otherwise I feel it gets more confusing if context windows are ordered differently.

That is a good question. I honestly haven't thought about that, but I would agree with you - that if we flip order, we should also display results in context in flipped order, so they read in the same way. @jessover9000 @davkal do you agree?

davkal commented:
> if we flip order, we should also display results in context in flipped order, so they read in the same way

Yup, 100%.

davkal commented:
> Should we add any visual feedback that the lines are switching order (e.g. icon in button with arrow up/down)? Or do you think it works without it?

It's mostly clear from looking at the result what the sort order is. Perhaps it's worth adding it as a title to the button so you could see it on demand. Another thing I noticed is that the flip takes a second or two on my machine, so I can imagine a button state that is makes it disabled for a short time, say 1s, makes it clear that something is happening. But that's not a blocker.

davkal commented:
I had one concern about sorting being in the render path, but I profiled it on my slow machine and sorting is roughly 1-3ms (and only triggered once flipped), so negligible and the grand scheme of log line rendering.

davkal reviewed (commented):
Works well!

Blocker:
- context order should be same

Minor:
- sorting in render path, perhaps memoize?
- sort order could be added to button title

Questions:
- sorting is done on `timeEpochMs`, could there be lines that only differ in `timeEpochNs`?

ivanahuckova committed:
Move SortOrder enum to grafana/data

ivanahuckova committed:
Update SortOrder in test

ivanahuckova committed:
Update context based on sort order of logs

ivanahuckova committed:
Update used  method for panel, update tests

ivanahuckova committed:
Rename prop  to logsSortOrder

ivanahuckova committed:
Memoize resuults

ivanahuckova committed:
Add title too button

ivanahuckova commented:
>Another question is what to do with log context? Should we flip order there too? 

✅ Done, order of context flips based on order of results.

>Perhaps it's worth adding it as a title to the button so you could see it on demand.

✅ Done, I have added the title `Change to descending order` or  `Change to ascending order`. @jessover9000 @davkal do you think that something like `Change to newest first` or `Change to newest last` would be more easier to understand? 

>Sorting in render path, perhaps memoize?

✅ Memoized

>Sorting is done on timeEpochMs, could there be lines that only differ in timeEpochNs?

✅ Sorting is done on timeEpochNs. timeEpochMs was used only in tests, but I have moved the sorting functions to grafana/data and now we are able to use it for grafana/ui components + tests (e.g. Logrows).

ivanahuckova committed:
Add disablinng of button for 1sec

ivanahuckova commented on packages/grafana-data/src/types/logs.ts:
```diff
@@ -33,6 +33,11 @@ export enum LogsMetaKind {
+  Descending = 'Descending',
+  Ascending = 'Ascending',
+}
+
```
moved from core to grafana/data so we can use it in LogRows that is in grafana/ui

ivanahuckova commented on packages/grafana-data/src/utils/logs.ts:
```diff
@@ -1,6 +1,6 @@
 import { countBy, chain, escapeRegExp } from 'lodash';
```
moved from core to grafana/data so we can use it in LogRows that is in grafana/ui

ivanahuckova replied:
I have moved `logs sorting functions` from grafana `core` to `grafana/data` so we can use it in `grafana/ui` components such as `LogsRow`.

ivanahuckova commented on packages/grafana-ui/src/components/Logs/LogRow.tsx:
```diff
@@ -206,11 +208,16 @@ class UnThemedLogRow extends PureComponent<Props, State> {
+            row={this.props.row}
+            getRowContext={this.props.getRowContext}
+            logsSortOrder={logsSortOrder}
+          >
```
we are passing logsSortOrder so we can provide correct context based on sort order.

ivanahuckova replied:
Passing `sort order` to `LogRowContextProvider` so we can flip order of context as well.

ivanahuckova commented on packages/grafana-ui/src/components/Logs/LogRowContextProvider.tsx:
```diff
@@ -58,69 +60,74 @@ export const getRowContexts = async (
 
   const results: Array<DataQueryResponse | DataQueryError> = await Promise.all(promises.map(p => p.catch(e => e)));
 
-  return {
```
Saving data and errors to variables before returning them.

ivanahuckova commented on packages/grafana-ui/src/components/Logs/LogRowContextProvider.tsx:
```diff
@@ -58,69 +60,74 @@ export const getRowContexts = async (
-      if (!errorResult.message) {
-        return '';
-      }
+    return logsSortOrder === LogsSortOrder.Ascending ? data.reverse() : data;
```
If order is Ascending (flipped), we need to reverse order of logs.

ivanahuckova replied:
If sort order if `ascending` (flipped), we have to reverse the order of logs in both - `before` and `after` arrays.

ivanahuckova commented on packages/grafana-ui/src/components/Logs/LogRowContextProvider.tsx:
```diff
@@ -58,69 +60,74 @@ export const getRowContexts = async (
+  return {
+    data: logsSortOrder === LogsSortOrder.Ascending ? data.reverse() : data,
+    errors: logsSortOrder === LogsSortOrder.Ascending ? errors.reverse() : errors,
   };
```
If order is Ascending (flipped), we need to reverse order of before and after arrays of logs.

ivanahuckova commented on public/sass/components/_panel_logs.scss:
```diff
@@ -12,12 +12,18 @@ $column-horizontal-spacing: 10px;
 
 .logs-panel-controls {
```
styling of Flip logs order button based on feedback from Jess.

ivanahuckova commented on public/app/plugins/panel/logs/types.ts:
```diff
@@ -1,8 +1,8 @@
-import { SortOrder } from 'app/core/utils/explore';
+import { LogsSortOrder } from '@grafana/data';
 
```
Change of SortOrder import as we have moved it to grafana/data.

ivanahuckova commented on public/app/plugins/panel/logs/module.tsx:
```diff
@@ -1,7 +1,6 @@
-import { PanelPlugin } from '@grafana/data';
```
Change of SortOrder import as we have moved it to grafana/data.

ivanahuckova commented on public/app/plugins/panel/logs/LogsPanel.tsx:
```diff
@@ -3,7 +3,6 @@ import { LogRows, CustomScrollbar } from '@grafana/ui';
 import { LogsDedupStrategy, PanelProps } from '@grafana/data';
```
Change of SortOrder import as we have moved it to grafana/data. Plus we are just passing order of logs and the ordering is processed in LogRows - the same way as in explore.

ivanahuckova commented on public/app/features/explore/utils/ResultProcessor.ts:
```diff
@@ -7,9 +7,10 @@ import {
   getDisplayProcessor,
```
Change of SortOrder import as we have moved it to grafana/data.

ivanahuckova commented on public/app/features/explore/state/reducers.ts:
```diff
@@ -12,6 +12,7 @@ import {
   toLegacyResponseData,
```
Change of SortOrder import as we have moved it to grafana/data.

ivanahuckova commented on public/app/features/explore/RichHistory/RichHistoryStarredTab.tsx:
```diff
@@ -8,9 +8,7 @@ import { RichHistoryQuery, ExploreId } from 'app/types/explore';
 // Utils
```
Change of SortOrder import as we have moved LogsSortOrder to grafana/data and moved SortOrder for Rich history to `app/core/utils/richHistory`.

ivanahuckova commented on public/app/features/explore/RichHistory/RichHistoryStarredTab.test.tsx:
```diff
@@ -1,7 +1,7 @@
 import React from 'react';
```
Change of SortOrder import as we have moved LogsSortOrder to grafana/data and moved SortOrder for Rich history to app/core/utils/richHistory.

ivanahuckova commented on public/app/features/explore/RichHistory/RichHistoryQueriesTab.tsx:
```diff
@@ -9,8 +9,8 @@ import { RichHistoryQuery, ExploreId } from 'app/types/explore';
 import { stylesFactory, useTheme } from '@grafana/ui';
```
Change of SortOrder import as we have moved LogsSortOrder to grafana/data and moved SortOrder for Rich history to app/core/utils/richHistory.

ivanahuckova commented on public/app/features/explore/RichHistory/RichHistoryQueriesTab.test.tsx:
```diff
@@ -1,7 +1,7 @@
 import React from 'react';
```
Change of SortOrder import as we have moved LogsSortOrder to grafana/data and moved SortOrder for Rich history to app/core/utils/richHistory.

ivanahuckova commented on public/app/features/explore/RichHistory/RichHistory.tsx:
```diff
@@ -2,8 +2,7 @@ import React, { PureComponent } from 'react';
 import { css } from 'emotion';
```
Change of SortOrder import as we have moved LogsSortOrder to grafana/data and moved SortOrder for Rich history to app/core/utils/richHistory.

ivanahuckova commented on public/app/features/explore/Logs.tsx:
```diff
@@ -205,23 +225,41 @@ export class Logs extends PureComponent<Props, State> {
+                ))}
+              </ToggleButtonGroup>
+            </div>
+            <button
```
We are using "old-style" button to keep the Explore UI consistent. In the future, will be changed to <Button />

ivanahuckova replied:
We are using old button to keep UI consistent.

ivanahuckova commented on public/app/core/utils/richHistory.ts:
```diff
@@ -5,7 +5,6 @@ import _ from 'lodash';
 import { DataQuery, DataSourceApi, dateTimeFormat, AppEvents, urlUtil, ExploreUrlState } from '@grafana/data';
```
Change of SortOrder import as we have moved LogsSortOrder to grafana/data and moved SortOrder for Rich history to app/core/utils/richHistory.

ivanahuckova commented on public/app/core/utils/richHistory.test.ts:
```diff
@@ -10,7 +10,7 @@ import {
   filterAndSortQueries,
```
Change of SortOrder import as we have moved LogsSortOrder to grafana/data and moved SortOrder for Rich history to app/core/utils/richHistory.

ivanahuckova commented on packages/grafana-data/src/utils/logs.test.ts:
```diff
@@ -1,4 +1,5 @@
-import { LogLevel } from '../types/logs';
```
As we have moved `LogsSortOrder`and `sorting functions` from core to grafana/data, we are moving the associated tests as well.

ivanahuckova replied:
I have moved `logs sorting functions` and associated tests from grafana `core` to `grafana/data` so we can use it in `grafana/ui` components such as `LogsRow`.

ivanahuckova commented on packages/grafana-ui/src/components/Logs/LogRowContextProvider.tsx:
```diff
@@ -1,4 +1,4 @@
-import { LogRowModel, toDataFrame, Field, FieldCache } from '@grafana/data';
```
In `LogRowContextProvider` we receive `sortOrder` as prop and we update the logs order in context based on it.

ivanahuckova commented on packages/grafana-ui/src/components/Logs/LogRows.test.tsx:
```diff
@@ -2,7 +2,7 @@ import React from 'react';
 import { range } from 'lodash';
```
Added test coverage for a new functionality - flipping the order of logs.

ivanahuckova commented on public/app/core/utils/explore.test.ts:
```diff
@@ -8,23 +8,12 @@ import {
   hasNonEmptyQuery,
```
Test for sorting were moved to `grafana/data`. Removing unused imports and tests.

ivanahuckova commented on public/app/core/utils/explore.ts:
```diff
@@ -14,7 +14,7 @@ import {
   IntervalValues,
```
`Sorting functions` were moved to `grafana/data`.

ivanahuckova commented on packages/grafana-data/src/types/logs.ts:
```diff
@@ -33,6 +33,11 @@ export enum LogsMetaKind {
   LabelsMap,
 }
 
+export enum LogsSortOrder {
```
I have renamed `SortOrder` to `LogsSortOrder` for more semantic and clear name and moved it to `grafana/data` so we can use it in `grafana/ui` components such as `LogsRow`.

ivanahuckova commented on packages/grafana-ui/src/components/Logs/LogRowContextProvider.tsx:
```diff
@@ -58,69 +60,74 @@ export const getRowContexts = async (
-      if (!dataResult.data) {
-        return [];
-      }
+  const data = results.map(result => {
```
Saving `data` and `errors` as variables instead of just returning them, so we can modify their order based on `logs sort order`

ivanahuckova commented on packages/grafana-ui/src/components/Logs/LogRowContextProvider.tsx:
```diff
@@ -58,69 +60,74 @@ export const getRowContexts = async (
+    return errorResult.message;
+  });
+
+  return {
```
If sort order if `ascending` (flipped), we have to reverse the order of `before` and `after` arrays.

ivanahuckova commented on packages/grafana-ui/src/components/Logs/LogRows.test.tsx:
```diff
@@ -93,10 +93,88 @@ describe('LogRows', () => {
 
     expect(wrapper.find(LogRow).length).toBe(100);
   });
+
```
Added test coverage for a new flipping order functionality.

ivanahuckova commented on packages/grafana-ui/src/components/Logs/LogRows.tsx:
```diff
@@ -139,6 +146,7 @@ class UnThemedLogRows extends PureComponent<Props, State> {
                   onClickFilterOutLabel={onClickFilterOutLabel}
                   getFieldLinks={getFieldLinks}
+                  logsSortOrder={logsSortOrder}
                 />
```
Passing `logsSortOrder` to `LogRow` so it can be passed on `LogsRowContextProvider`.

ivanahuckova commented on public/app/core/logs_model.ts:
```diff
@@ -28,10 +28,11 @@ import {
 import { getThemeColor } from 'app/core/utils/colors';
 
-import { sortInAscendingOrder, deduplicateLogRowsById } from 'app/core/utils/explore';
+import { deduplicateLogRowsById } from 'app/core/utils/explore';
```
Changed import as `sorting functions` were moved to `grafana/data`.

ivanahuckova commented on public/app/core/utils/richHistory.test.ts:
```diff
@@ -10,7 +10,7 @@ import {
 } from './richHistory';
 import store from 'app/core/store';
-import { SortOrder } from './explore';
+import { SortOrder } from './richHistory';
```
Changed import as `Sort order` for rich history was moved to `./richHistory`.

ivanahuckova commented on public/app/core/utils/richHistory.ts:
```diff
@@ -5,7 +5,6 @@ import _ from 'lodash';
 import { DataQuery, DataSourceApi, dateTimeFormat, AppEvents, urlUtil, ExploreUrlState } from '@grafana/data';
 import appEvents from 'app/core/app_events';
 import store from 'app/core/store';
-import { SortOrder } from './explore';
```
`Sort order` for rich history was moved to `./richHistory` and `Logs sort order` was moved to `grafana/data`.

ivanahuckova commented on public/app/features/explore/RichHistory/RichHistory.tsx:
```diff
@@ -2,8 +2,7 @@ import React, { PureComponent } from 'react';
 //Services & Utils
-import { SortOrder } from 'app/core/utils/explore';
-import { RICH_HISTORY_SETTING_KEYS } from 'app/core/utils/richHistory';
+import { RICH_HISTORY_SETTING_KEYS, SortOrder } from 'app/core/utils/richHistory';
```
`Sort order` for rich history was moved to `./richHistory`.

ivanahuckova commented on public/app/features/explore/RichHistory/RichHistoryQueriesTab.test.tsx:
```diff
@@ -1,7 +1,7 @@
 import { mount } from 'enzyme';
 import { ExploreId } from '../../../types/explore';
-import { SortOrder } from 'app/core/utils/explore';
+import { SortOrder } from 'app/core/utils/richHistory';
```
Sort order for rich history was moved to ./richHistory.

ivanahuckova commented on public/app/features/explore/RichHistory/RichHistoryQueriesTab.tsx:
```diff
@@ -9,8 +9,8 @@ import { RichHistoryQuery, ExploreId } from 'app/types/explore';
 
-import { SortOrder } from 'app/core/utils/explore';
 import {
+  SortOrder,
```
Sort order for rich history was moved to ./richHistory.

ivanahuckova commented on public/app/features/explore/RichHistory/RichHistoryStarredTab.test.tsx:
```diff
@@ -1,7 +1,7 @@
 import { mount } from 'enzyme';
 import { ExploreId } from '../../../types/explore';
-import { SortOrder } from 'app/core/utils/explore';
+import { SortOrder } from 'app/core/utils/richHistory';
```
Sort order for rich history was moved to ./richHistory.

ivanahuckova commented on public/app/features/explore/RichHistory/RichHistoryStarredTab.tsx:
```diff
@@ -8,9 +8,7 @@ import { RichHistoryQuery, ExploreId } from 'app/types/explore';
-
-import { SortOrder } from '../../../core/utils/explore';
-import { filterAndSortQueries, createDatasourcesList } from '../../../core/utils/richHistory';
+import { filterAndSortQueries, createDatasourcesList, SortOrder } from 'app/core/utils/richHistory';
```
Sort order for rich history was moved to ./richHistory.

ivanahuckova commented on public/app/features/explore/state/reducers.ts:
```diff
@@ -12,6 +12,7 @@ import {
   toLegacyResponseData,
   ExploreMode,
   LogsDedupStrategy,
+  sortLogsResult,
```
`sortLogsResult` was moved to `grafana/data`.

ivanahuckova commented on public/app/features/explore/utils/ResultProcessor.ts:
```diff
@@ -7,9 +7,10 @@ import {
   PreferredVisualisationType,
   standardTransformers,
+  sortLogsResult,
 } from '@grafana/data';
```
`sortLogsResult` was moved to `grafana/data`.

ivanahuckova commented on public/app/plugins/panel/logs/LogsPanel.tsx:
```diff
@@ -22,12 +21,11 @@ export const LogsPanel: React.FunctionComponent<LogsPanelProps> = ({
 
   const newResults = data ? dataFrameToLogsModel(data.series, data.request?.intervalMs, timeZone) : null;
-  const sortedNewResults = sortLogsResult(newResults, sortOrder);
 
```
Instead of sorting here, we are passing the `sort order` to `LogsRow` and using the same functionality as in Explore to sort the order of logs.

ivanahuckova commented on public/app/plugins/panel/logs/module.tsx:
```diff
@@ -1,7 +1,6 @@
-import { PanelPlugin } from '@grafana/data';
+import { PanelPlugin, LogsSortOrder } from '@grafana/data';
```
`LogsSortOrder `was renamed and moved to `grafana/data`.

ivanahuckova commented on public/app/plugins/panel/logs/types.ts:
```diff
@@ -1,8 +1,8 @@
-import { SortOrder } from 'app/core/utils/explore';
+import { LogsSortOrder } from '@grafana/data';
```
LogsSortOrder was renamed and moved to grafana/data.

ivanahuckova commented on public/sass/components/_panel_logs.scss:
```diff
@@ -12,12 +12,18 @@ $column-horizontal-spacing: 10px;
 
```
`css` updates based on design.

jessover9000 commented:
> do you think that something like Change to newest first or Change to newest last would be more easier to understand?

I have a degree in linguistics and still I always have to think about "ascending" and "descending" as in, which is which again. 🤪  So I'm sure "newest first" and "oldest first" will reduce cognitive load! (I would use "oldest first" and not "newest last")

ivanahuckova committed:
Update wordiing

ivanahuckova commented:
I agree @jessover9000! Thanks for the advice. Updated to `Change to newest first` and `Change to oldest first`.

zoltanbedi commented on packages/grafana-ui/src/components/Logs/LogRows.test.tsx:
```diff
@@ -93,10 +93,88 @@ describe('LogRows', () => {
+
+  it('renders asc ordered rows if order and function supplied', () => {
+    const rows: LogRowModel[] = [
+      makeLog({ uid: '1', timeEpochMs: 1 }),
```
I think it would make sense to change the order of this array to make sure that the ordering works well.

ivanahuckova replied:
Yeah! Great point. Updated in https://github.com/grafana/grafana/pull/26669/commits/8cc825a76add3576a87e6369064a2ee80d99b1d7

zoltanbedi reviewed (approved):
Good job on this! 🧙‍♀️

ivanahuckova committed:
Update packages/grafana-data/src/utils/logs.ts

Co-authored-by: Zoltán Bedi <zoltan.bedi@gmail.com>

ivanahuckova committed:
Update packages/grafana-data/src/utils/logs.ts

Co-authored-by: Zoltán Bedi <zoltan.bedi@gmail.com>

ivanahuckova committed:
Update test by reordering logs

ivanahuckova committed:
Merge remote-tracking branch 'origin/master' into ivana/sort-log-results

davkal commented on public/app/features/explore/Logs.tsx:
```diff
@@ -72,13 +74,31 @@ interface State {
+  onChangeLogsSortOrder = () => {
+    this.setState({ disabledSortButton: true });
+    // we are using setTimeout here to make sure that disabled button is rendered before the rendering of reordered logs
+    setTimeout(() => {
```
To be really clean, you'd track both timeouts in component variables and clear them before component unmount.

davkal reviewed (approved):
Works well, and code looks good. 

Small nit:

- tracking settimeout inside the component is missing
- the button text after click could read "Flipping...", and then reset once enabled again.

ivanahuckova committed:
Clear timers, add button flipping title

ivanahuckova commented:
Both nits fixed in https://github.com/grafana/grafana/pull/26669/commits/33c38176de659a10caeae9b94bf4ca58b49e66ef
